### PR TITLE
[WIP] Added Prototype for Exporting RV

### PR DESCRIPTION
**Pull request by ritesh99rakesh**
#### References to other Issues or PRs

Refer to the discussion: https://github.com/sympy/sympy/issues/16942

#### Brief description of what is fixed or changed
Added `_sample_numpy` and `_sample_scipy` for `Normal` distribution under
`sympy/stats/crv_types.py`

#### Other comments
Suggestions required.
Ping @Upabjojr @sidhantnagpal @czgdp1807 

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by ritesh99rakesh** [comment]
@Upabjojr I have added the sample function using numpy and scipy for normal distribution. But as you can see, there is a lot of code duplication.
Therefore, I feel that we should discuss the exact nature of the functions and place where we should keep them.
Please review.

File: sympy/stats/crv_types.py [hunk 1]
```diff
@@ -2320,6 +2320,32 @@ def pdf(self, x):
     def sample(self):
         return random.normalvariate(self.mean, self.std)
 
+    def _sample_numpy(self, **kwargs):
+        mean, std = self.mean, self.std
+        if isinstance(mean, Symbol) or isinstance(std, Symbol):
```
Comment (czgdp1807) [inline sympy/stats/crv_types.py]: Just a suggestion for the future work, use, something like, `mean.has(Symbol) or std.has(Symbol)`. No need to make changes now, as this is a prototype.

File: sympy/stats/crv_types.py [hunk 2]
```diff
@@ -2320,6 +2320,32 @@ def pdf(self, x):
...
+            raise ValueError('Cannot sample from Symbolic parameters.')
+
+        numpy = import_module('numpy')
+        if numpy:
+            mean, std = float(mean), float(std)
```
Comment (czgdp1807) [inline sympy/stats/crv_types.py]: Casting may invite errors in computation. Is there any other method for going from `SymPy` objects to `float` objects.

Comment (czgdp1807) [reply sympy/stats/crv_types.py]: Like, using `args` and all that stuff. Please try to remove casting, if possible.

Comment (ritesh99rakesh) [reply sympy/stats/crv_types.py]: I don't have much idea in this area. Maybe @sidhantnagpal clarify this.

Comment (czgdp1807) [reply sympy/stats/crv_types.py]: May be there isn't a way. I tried it out. Comment if you can find out any, otherwise, it's fine to do casting. I think, `SymPy` has its own method like, `__float__` which can avoid errors. Can you check?

**Comment by sidhantnagpal** [comment]
Much of the duplication can be removed if you factor out the common logic (like ensuring numeric `mu`, `sigma` and `import_module` checks) and move it to `sample` method (which could itself be moved to superclass).

This is assuming `_sample_*` methods are planned to be accessed only through `sample`.

**Comment by ritesh99rakesh** [comment]
> This is assuming `_sample_*` methods are planned to be accessed only through `sample`.

Yes @sidhantnagpal I believe that `_sample_*` methods are supposed to called internally only through `sample`. @Upabjojr what do you think?

File: sympy/stats/crv_types.py (hunk 2 above)
Comment (ritesh99rakesh) [reply sympy/stats/crv_types.py]: `__float__` method internally uses `float` to convert to float.
```python
def __float__(self):
    # Don't bother testing if it's a number; if it's not this is going
    # to fail, and if it is we still need to check that it evalf'ed to
    # a number.
    result = self.evalf()
    if result.is_Number:
        return float(result)
    if result.is_number and result.as_real_imag()[1]:
        raise TypeError("can't convert complex to float")
    raise TypeError("can't convert expression to float")
```
So, I think above is fine.

**Comment by ritesh99rakesh** [comment]
@sidhantnagpal what do you think about this approach?

File: sympy/stats/crv_types.py [hunk 3]
```diff
@@ -2317,8 +2317,33 @@ def check(mean, std):
     def pdf(self, x):
         return exp(-(x - self.mean)**2 / (2*self.std**2)) / (sqrt(2*pi)*self.std)
 
-    def sample(self):
-        return random.normalvariate(self.mean, self.std)
+    def sample(self, **kwargs):
+        mean, std = self.mean, self.std
+        print("len: ", len(kwargs))
```
Comment (sidhantnagpal) [inline sympy/stats/crv_types.py]: This should be removed.

File: sympy/stats/crv_types.py [hunk 4]
```diff
@@ -2317,8 +2317,33 @@ def check(mean, std):
...
+        if len(kwargs) == 0 or mean.has(Symbol) or std.has(Symbol):
+            return random.normalvariate(self.mean, self.std)
+        objs = {}
+        mean, std, size = float(mean), float(std), kwargs.get('size', 1)
+        if kwargs.get('numpy', False):
+            objs['numpy'] = _sample_numpy(mean, std, size)
```
Comment (sidhantnagpal) [inline sympy/stats/crv_types.py]: Don't you intend to return these?

Comment (ritesh99rakesh) [reply sympy/stats/crv_types.py]: Yes. I will make the correction.

File: sympy/stats/crv_types.py [hunk 5]
```diff
@@ -2317,8 +2317,33 @@ def check(mean, std):
...
+        if kwargs.get('scipy', False):
+            objs['scipy'] = _sample_scipy(mean, std, size)
+
+    def _sample_numpy(mean, std, size):
+        numpy = import_module('numpy')
+        if numpy:
+            return numpy.random.normal(mean, std, size)
```
Comment (Upabjojr) [inline sympy/stats/crv_types.py]: this method should just return, no module import.

File: sympy/stats/crv_types.py [hunk 6]
```diff
@@ -2317,8 +2317,33 @@ def check(mean, std):
     def pdf(self, x):
         return exp(-(x - self.mean)**2 / (2*self.std**2)) / (sqrt(2*pi)*self.std)
 
-    def sample(self):
-        return random.normalvariate(self.mean, self.std)
+    def sample(self, **kwargs):
```
Comment (Upabjojr) [inline sympy/stats/crv_types.py]: Why is this here? It should belong to the superclass.

**Comment by Upabjojr** [comment]
> I believe that `_sample_*` methods are supposed to called internally only through `sample`. @Upabjojr what do you think?

Yes, that's the purpose of those methods.

File: sympy/stats/crv_types.py (hunk 4 above)
Comment (Upabjojr) [reply sympy/stats/crv_types.py]: If you are calling a class method, you should prepend `self.` to the method name.

**Comment by Upabjojr** [comment]
Where are the tests? I had a glance at your code and it really looks like it cannot work in the current state.

**Comment by Upabjojr** [comment]
@ritesh99rakesh please make this PR work. I would recommend transforming the currently existing `sample` methods into `_sample_python`, then define `sample` in the superclass so that it (temporarily) calls `_sample_python`, with the existing code.

As soon as the other methods are ready we can start adding a dispatcher for the other methods.

Please refrain temporarily from writing a dispatcher (e.g. to choose the methods). We will do it later, when the sampler are working.

Can you do this by tomorrow?

**Comment by ritesh99rakesh** [comment]
> Can you do this by tomorrow?

Yes, I will be able to complete it by today and will ask for your suggestions on the code.

File: sympy/stats/crv_types.py (hunk 2 above)
Comment (czgdp1807) [reply sympy/stats/crv_types.py]: Resolve this conversation.

**Comment by ritesh99rakesh** [comment]
@Upabjojr Following changes were made:
1. Added `_sample_python` to `NormalDistribution`.
2. Removed `_inverse_cdf_expression` function since we do not require it to find the sample.

As you can see that `_sample_python` in itself is not much dependent upon `NormalDistribution` and could be made generic by shifting the code to `sample` under `stats/crv.py`.

I look forward to your suggestions.

**Comment by Upabjojr** [comment]
>     2\. Removed `_inverse_cdf_expression` function since we do not require it to find the sample.

Why are you removing working code?!?!?!?!?

I just asked you to move the current `sample` function into `_sample_python`. Do not remove anything.

**Comment by Upabjojr** [comment]
We are going to add extra sampling functions (like `_sample_scipy`, `_sample_pymc3`, `_sample_stan`, etc...)

**Comment by ritesh99rakesh** [comment]
> I just asked you to move the current `sample` function into `_sample_python`. Do not remove anything.

Ok @Upabjojr! I will keep that in mind and will restore this.

File: sympy/stats/crv.py [hunk 7]
```diff
@@ -169,15 +169,15 @@ def __new__(cls, *args):
     def check(*args):
         pass
 
-    def sample(self):
+    def sample(self, **kwargs):
         """ A random realization from the distribution """
+        return self._sample_python(**kwargs)
         icdf = self._inverse_cdf_expression()
         return icdf(random.uniform(0, 1))
```
Comment (sidhantnagpal) [inline sympy/stats/crv.py]: This code is not reachable due to early return.

Comment (ritesh99rakesh) [reply sympy/stats/crv.py]: Yes, I know. I have added my code without deleting the previous code so that we can finalise the prototype.
Also, could you tell me your review about returning the `objs` in `NormalDistribution` under `sympy/stats/crv_types.py`? The current return type is `object`. Is there a better way?

File: sympy/stats/crv_types.py [hunk 8]
```diff
@@ -2339,9 +2339,60 @@ def check(mean, std):
     def pdf(self, x):
         return exp(-(x - self.mean)**2 / (2*self.std**2)) / (sqrt(2*pi)*self.std)
 
-    def sample(self):
+    def _sample_python(self, **kwargs):
```
Comment (Upabjojr) [inline sympy/stats/crv_types.py]: This method is supposed to be the only `sample` method, just that.

File: sympy/stats/rv.py [hunk 9]
```diff
@@ -369,7 +369,7 @@ def domain(self):
     def density(self):
         raise NotImplementedError("Density not available for ProductSpaces")
 
-    def sample(self):
+    def sample(self, **kwargs):
```
Comment (Upabjojr) [inline sympy/stats/rv.py]: why `**kwargs`?

**Comment by Upabjojr** [comment]
> Where are the tests? I had a glance at your code and it really looks like it cannot work in the current state.

I still see no tests. Where do you test the new methods you have added?

**Comment by Upabjojr** [comment]
I'm closing this PR as it's becoming quite a mess.

Please start over. Let's have two PRs:

- addition of new methods (e.g. `_sample_pymc3`, `_sample_scipy`). Just that, no dispatchers for now.
- moving `sample` to `_samply_python`, just that.

**Comment by Upabjojr** [comment]
And espacially, please add some tests to every PR.
